nulib/meadow
# Store preservation files in S3 by id rather than sha256

# Summary 

We don't want our preservation storage location tied to a specific algorithm so we are changing the naming in the preservation bucket to be based on Meadow's file set id. 

# Specific Changes in this PR

- CopyFileToPreservation action works differently
  - Store preservation files in S3 by id rather than `sha256`
  - write object tags in addition to object metadata
- Switch to locally forked `ex_aws_s3` in order to support replacing tags during `CopyFileToPreservation` action

# Version Bump

# Steps to Test
- Ingest a work with a file set, and check preservation location is the pairtree'd id and that the 2 checksum tags exist on the object (along with the metadata we've been creating already)

# :rocket: Deployment Notes

- Backward compatible API changes
- Backwards-incompatible API changes
  - [x] Requires coordination with S3 Batch Operation renaming task: https://github.com/nulib/s3-batch-renamer-lambda
  - [x] Requires manually updating preservation location in all file set database records, post-migration (after this PR is merged)
```
Repo.transaction(
  fn ->
    from(f in Meadow.Data.Schemas.FileSet)
    |> Repo.stream()
    |> Stream.each(fn file_set ->
        file_set
        |> Meadow.Data.Schemas.FileSet.changeset(%{core_metadata: %{location: Meadow.Data.FileSets.preservation_location(file_set)}})
        |> Repo.update()
      end)
    |> Stream.run()
  end,
  timeout: :infinity
)
```

# Tested/Verified
